File: LeetCode/go/LC0045_JumpGameII/jump_game_ii.go
package jump_game_ii

func jump(nums []int) int {
	length := len(nums)
	end := 0
	maxPos := 0
	steps := 0
	for i := 0; i < length-1; i++ {
		maxPos = max(maxPos, i+nums[i])
		if i == end {
			end = maxPos
			steps++
		}
	}
	return steps
}

File: LeetCode/java/LC0045_JumpGameII/JumpGameII.java
package LC0045_JumpGameII;

/**
 * <a href="https://leetcode.cn/problems/jump-game-ii/">Jump Game II</a>
 * 贪心；数组；动态规划
 */
public class JumpGameII {
    public static void main(String[] args) {
        Solution solution = new Solution();
        System.out.println(solution.jump(new int[]{2, 3, 1, 1, 4}));
    }
}

class Solution {
    public int jump(int[] nums) {
        int end = 0;
        int maxPos = 0;
        int steps = 0;
        for (int i = 0; i < nums.length - 1; i++) {
            maxPos = Math.max(maxPos, i + nums[i]);
            if (i == end) {
                end = maxPos;
                steps++;
            }
        }
        return steps;
    }
}
